Introduce "order book not crossed" invariant (for fuzzing only)

rokopt opened the pull request:
# Introduce "order book not crossed" invariant

Resolves #2417 

This PR introduces a new invariant which keeps track of the state of the order book as operations are applied, and asserts that it is not crossed.

There is currently no mechanism by which an invariant can be informed that an operation application for which its `checkOnOperationApply()` method had been called is being rolled back.  Hence, we can not in general support an invariant such as this which maintains state across method calls.  But we can when fuzzing, where setup operations are never rolled back and where we can explicitly tell invariants when to snapshot state (that is, after setup completes) and when to reset it to that state (after each fuzz test).

Therefore, this invariant is compiled in when fuzzing is configured, and not otherwise.

This PR updates #2210 to current code, adapting to some changes such as the introduction of `InternalLedgerEntry` and the rearranging of the fuzzing code, and adds support for passive offers.

# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v8.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

rokopt commented:
This invariant _does_ appear to have a major effect on fuzzer execution speed.  Metrics after 20 minutes in current code without this PR:

![master-without-order-book-not-crossed-invariant-for-20-min](https://user-images.githubusercontent.com/17652016/110737886-2a2bb980-81e3-11eb-9684-0d88ee7960a4.png)

And with this PR:

![order-book-not-crossed-invariant-present-for-20-min](https://user-images.githubusercontent.com/17652016/110737903-30ba3100-81e3-11eb-8370-e84c805fd45c.png)

MonsieurNicolas commented on src/invariant/OrderBookIsNotCrossed.cpp:
```diff
@@ -0,0 +1,205 @@
+void
+OrderBookIsNotCrossed::deleteFromOrderBook(OfferEntry const& oe)
+{
+    mOrderBook[oe.selling][oe.buying].erase(oe);
```
Why isn't `mOrderbook` an `std::unordered_map<std::pair<Asset, Asset>, Orders>` ? All places I see that we're always looking things up with both assets fixed so we're just making things slower for no reason by nesting them.

rokopt replied:
You're right -- that made several passes shorter.

MonsieurNicolas commented on src/invariant/test/OrderBookIsNotCrossedTests.cpp:
```diff
@@ -0,0 +1,333 @@
+#ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
```
see me other comment: I'm pretty sure we can always test that the invariant works, regardless of fuzzing enabled or not

MonsieurNicolas replied:
(otherwise this test case will never be run anywhere)

MonsieurNicolas replied:
wow that was a jar jar binks comment from my part.
Meesa resolve this comment now!

MonsieurNicolas commented on src/main/ApplicationImpl.cpp:
```diff
@@ -177,6 +181,9 @@ ApplicationImpl::initialize(bool createNewDB)
     LiabilitiesMatchOffers::registerInvariant(*this);
     SponsorshipCountIsValid::registerInvariant(*this);
+#ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
+    OrderBookIsNotCrossed::registerInvariant(*this);
```
as this invariant does not work in the general case (from what I understand), we should probably register it only where we need it

rokopt replied:
I think I've managed that with the `registerAndEnableInvariant()` mechanism.

MonsieurNicolas commented on src/invariant/test/OrderBookIsNotCrossedTests.cpp:
```diff
@@ -0,0 +1,333 @@
+
+// convert a single offer into an offer operation
+Operation
+opFromLedgerEntries(LedgerEntry const& le)
```
I don't think those two function are needed: invariants don't care about operations

rokopt replied:
You're right, they don't -- the only loss from removing this (and the new `store()` parameter mentioned above) is that the `InvariantManager` won't report any interesting operation from an order-book-not-crossed-invariant test failure, and we'll be able to tell what the operation was by which test failed anyway, so there doesn't seem much point in having those functions.

rokopt replied:
I've removed them.

MonsieurNicolas commented on src/invariant/test/OrderBookIsNotCrossedTests.cpp:
```diff
@@ -0,0 +1,333 @@
+    }
+    else
+    {
+        updates = makeUpdateList(current, previous);
```
we should probably add a `REQUIRE` inside that version of `makeUpdateList` that ensures that keys are matching 1:1 (I know that code was already there, but it's super error prone especially in this context)

rokopt replied:
Added.

MonsieurNicolas commented on src/invariant/test/OrderBookIsNotCrossedTests.cpp:
```diff
@@ -0,0 +1,333 @@
+#include "test/TxTests.h"
+#include "test/test.h"
+
+using namespace stellar;
```
we normally prefer wrapping the whole block with `namespace stellar {`

rokopt replied:
Done.

MonsieurNicolas commented on src/invariant/test/OrderBookIsNotCrossedTests.cpp:
```diff
@@ -0,0 +1,333 @@
+}
+
+void
+applyCheck(Application& app, std::vector<LedgerEntry> const& current,
```
all those helper functions should be in the anonymous namespace as they're likely going to clash

rokopt replied:
Agreed, done.

MonsieurNicolas commented on src/invariant/test/OrderBookIsNotCrossedTests.cpp:
```diff
@@ -0,0 +1,333 @@
+    auto app = createTestApplication(clock, getTestConfig(0));
+    LedgerTxn ltxroot{app->getLedgerTxnRoot()};
+
+    auto const& invariant = std::make_shared<OrderBookIsNotCrossed>();
```
reference?

rokopt replied:
Fixed.

MonsieurNicolas commented on src/invariant/test/OrderBookIsNotCrossedTests.cpp:
```diff
@@ -0,0 +1,333 @@
+    // A against B
+    //     offer1: 3 A @  3/2 (B/A)
+    //     offer2: 2 A @  5/3 (B/A)
+    auto const& offer1 = createOffer(cur1, cur2, 3, Price{3, 2});
```
reference again?

MonsieurNicolas replied:
this file is filled with bogus refs to temp variable, please fix

rokopt replied:
Wow, those were weird.  I've gotten rid of all the ones I spotted.

sisuresh commented on src/invariant/OrderBookIsNotCrossed.h:
```diff
@@ -0,0 +1,84 @@
+#ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
+#pragma once
+
+// Copyright 2019 Stellar Development Foundation and contributors. Licensed
```
Year should be 2021. This comment applies to the two other files you added as well.

rokopt replied:
Done.

rokopt commented:
Alright, I've gone over the inherited files here and made a bunch of changes:

- The part I found trickiest was just getting the code to build and run in as many situations as possible!  We want it to build any time we have `BUILD_TESTS` configured, and to be able to run the invariant's own unit tests in that case, to minimize code rot.  And we want to build and use it when fuzzing is configured.  But we don't want the invariant to be active during any _other_ unit tests, because it maintains state across calls to `checkOnOperationApply()` and the code base in general does not support invariants that do that.  What I ended up doing was not registering the invariant in general because test configs enable `.*`, so if the invariant were registered, then it would be enabled during tests by default.  Instead, I exported a `registerAndEnableInvariant()`, unlike most invariants which export a `registerInvariant()` and let the config manager decide which to enable.  Then only those callers that explicitly know how to handle the order-book invariant's maintaining state can call `registerAndEnableInvariant()`.  And two callers do so:  the fuzzer and the invariant's own unit tests.  The invariant's own unit tests also have to disable _other_ invariants, because they modify ledger entries directly, which can break some other invariants.  (That part is something that several other invariant tests do as well -- they set `INVARIANT_CHECKS` to just themselves.)
- Rebased on master, which in particular includes the best-offer debugging code, with which there were a couple small conflicts.
- Improved the passive-offer-crossing-detection algorithm to break early if possible.
- Removed some weird C++ such as the references to temporaries and unnecessary `at()`s.
- Removed `using namespace` and added conventional namespace use.
- Added an assertion to `makeUpdateList()`.
- Deferred copying `mRolledBackOrderBook` as late as we can.
- Made `extractAssets()` clearer (I hope).
- Removed some changes which aren't necessary anymore after the above changes.

rokopt commented on src/invariant/OrderBookIsNotCrossed.cpp:
```diff
@@ -0,0 +1,205 @@
+                   (offer.flags & PASSIVE_FLAG) == 0;
+        };
+
+        auto nonPassiveAskAtSamePrice =
```
Done.

rokopt commented on src/invariant/OrderBookIsNotCrossed.cpp:
```diff
@@ -0,0 +1,205 @@
+                                   std::placeholders::_1));
+        auto nonPassiveBidAtSamePrice =
+            std::find_if(bids.cbegin(), bids.cend(),
+                         std::bind(findEqualPriceNonPassive, 1 / lowestAsk,
```
Actually not -- this is the case where `lowestAsk==highestBid`, so it had better not be a requirement that `1/lowestAsk==highestBid` as well, or else we've seriously narrowed down what `lowestAsk` is allowed to be.  ;-)

rokopt commented on src/invariant/OrderBookIsNotCrossed.cpp:
```diff
@@ -0,0 +1,205 @@
+        if (entry.first.type() == InternalLedgerEntryType::LEDGER_ENTRY &&
+            entry.first.ledgerKey().type() == OFFER && entry.second.current)
+        {
+            auto const& offer =
```
I hope I've made it more readable now by ordering the pairs as you suggest.

MonsieurNicolas replied:
yes, thank you!

rokopt commented on src/invariant/OrderBookIsNotCrossed.cpp:
```diff
@@ -201,5 +201,17 @@ OrderBookIsNotCrossed::checkOnOperationApply(Operation const& operation,
+void
+OrderBookIsNotCrossed::resetForFuzzer()
+{
+    mOrderBook = mRolledBackOrderBook;
```
Good idea, done.

MonsieurNicolas replied:
I see that you reverted your change. From the looks of it, it was just not correct.
you need to always `clear` in `resetForFuzzer` (for stability) and CoW

rokopt commented on src/test/FuzzerImpl.cpp:
```diff
@@ -743,6 +743,7 @@ resetTxInternalState(Application& app)
 // reset caches to clear persistent state
 #ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
     app.getLedgerTxnRoot().resetForFuzzer();
+    app.getInvariantManager().resetForFuzzer();
```
Done.

rokopt commented on src/invariant/test/InvariantTestUtils.cpp:
```diff
@@ -57,7 +57,7 @@ generateOffer(Asset const& selling, Asset const& buying, int64_t amount,
 bool
 store(Application& app, UpdateList const& apply, AbstractLedgerTxn* ltxPtr,
-      OperationResult const* resPtr)
+      OperationResult const* resPtr, Operation const* opPtr)
```
It had been used in the previous PR, but other changes in this one have allowed it to be removed.

jonjove commented on src/invariant/OrderBookIsNotCrossed.cpp:
```diff
@@ -0,0 +1,225 @@
+        auto nonPassiveBidAtSamePrice =
+            existsEqualPriceNonPassive(1 / lowestAsk, bids, bid);
+
+        if (nonPassiveAskAtSamePrice && nonPassiveBidAtSamePrice)
```
Based on all of my comments here, I believe that there is actually an O(1) way to perform this check. Your current implementation is O(n) so this would be a big improvement.

rokopt replied:
My comment above (https://github.com/stellar/stellar-core/pull/2959#discussion_r595418524) explains why I currently do not see how we can avoid searching a list of offers at a price that exactly equally crosses some pre-existing offer to find out whether any of them is not passive, if we are to catch a bug where there is such an offer.  (Unless we were to check and update state with each individual `Operation` in `checkOnOperationApply`, rather than extracting all the asset pairs and batching the updates and then doing a single check, but I would expect that to be slowing down more common cases to speed up this rarer one.)  I might be misreading the semantics of passive offers, though!

jonjove replied:
Despite your clarifications about how this implementation works, I still believe this can be done with O(1) complexity.

rokopt replied:
I eventually caught on to Jon's clever idea here:  extend `OfferEntryCmp` to sort non-passive orders before passive orders at the same price.  Then we know that if there's _any_ non-passive order in the list of offers at the same price, then one of the non-passive orders will be at the front.  That allows us to check only the fronts of each list.

jonjove commented on src/invariant/OrderBookIsNotCrossed.cpp:
```diff
@@ -0,0 +1,225 @@
+void
+OrderBookIsNotCrossed::resetForFuzzer()
+{
+    mRollBackBeforeNextCheck = true;
```
I'm not 100% sure I understand the reasoning for this implementation. You could do `mOrderBook = mRolledBackOrderBook` here, right? I suppose I can imagine some reasons for not doing that. What was your reasoning? You probably want to put it in a comment if it's important.

rokopt replied:
That was Nicolas's idea, and, I think, a good one: https://github.com/stellar/stellar-core/pull/2959#discussion_r592848513

The point is that some fuzz operation sets won't involve orders, and in those cases we call `OrderBookIsNotCrossed::checkOnOperationApply()` but not `OrderBookIsNotCrossed::check()`.  In such a case, we can skip the copy and allow the invariant's representation of the order book to be stale until we actually use it, in `OrderBookIsNotCrossed::check()`.  I'll add a comment to that effect.

jonjove replied:
That's the same reason I came up with.

jonjove commented on src/invariant/OrderBookIsNotCrossed.cpp:
```diff
@@ -0,0 +1,225 @@
+}
+
+void
+OrderBookIsNotCrossed::deleteFromOrderBook(OfferEntry const& oe)
```
Nitpick: are these functions really adding any value? They're both one liners that are easy to understand, and they have a single caller.

rokopt replied:
Agreed, I'll inline those.

jonjove commented on src/invariant/OrderBookIsNotCrossed.h:
```diff
@@ -0,0 +1,99 @@
+
+using Orders = std::set<OfferEntry, OfferEntryCmp>;
+using OrderBook = std::unordered_map<AssetPair, Orders, AssetPairHash>;
+using AssetPairSet = std::set<std::pair<Asset, Asset>>;
```
Why not `unordered_set`?

rokopt replied:
I have actually tried it at a few different stages of development, but so far I haven't found a point where fuzzing was clearly faster with `unordered_set`, and at times it's been slower.  There's a lot of variation on my laptop so it's hard to be sure, and I was surprised at first, but it's actually not implausible that in this case `set` might be faster:  this is a fuzzing-only invariant and when fuzzing there are only ever a very small number of orders for a given asset pair, so we might be doing one or zero comparisons typically with `set`, and with `unordered_set` we always have to do one hash.

If we ever make this invariant workable on production-size order books, then I would guess that `unordered_set` would become clearly better, but the current version of this invariant isn't.  :)

jonjove replied:
That's a good reason.

jonjove reviewed (changes requested):
I didn't review the tests yet because I think the implementation is going to change. You should add tests for some of the interesting floating-point edge cases I pointed out.

rokopt commented on src/invariant/OrderBookIsNotCrossed.cpp:
```diff
@@ -0,0 +1,225 @@
+
+        OfferEntry ask, bid;
+        auto nonPassiveAskAtSamePrice =
+            existsEqualPriceNonPassive(lowestAsk, asks, ask);
```
Here's what I think is true: there is an ask with price `lowestAsk`, but it could be passive, and if it is, then it doesn't indicate by itself that the order book is crossed (or perhaps it's more accurate to say, crossed in a way that we don't intend?  We intend, as I understand it, for a passive ask/bid to be able to be placed into the book even if there is an existing bid/ask at the same price).  However, if there is any other ask at the same price which isn't passive, and a bid at the same price which isn't passive, then we know that some non-passive ask/bid arrived at a time when another bid/ask at the same price was already present, so the new, non-passive offer should take taken the existing one; we must have a bug.  Is that right?

jonjove replied:
Okay this works. Comment highly recommended because that is very strange looking.

rokopt replied:
In light of discussion below, the implementation can be made faster.  The replacement will also need comments since it's a bit tricky too.  :)

rokopt commented on src/test/FuzzerImpl.cpp:
```diff
@@ -1327,6 +1328,10 @@ TransactionFuzzer::initialize()
     ltxOuter.commit();
+
+#ifdef BUILD_TESTS
+    mApp->getInvariantManager().snapshotForFuzzer();
```
The fuzzer already uses `txtest`, and `getTestConfig()`, and `TestAccount`, so I think we committed to that a while ago.  We could revisit that if fuzzing without testing infrastructure has significant value, of course, but it's not clear to me that it does, and there would be a lot to change besides `snapshotForFuzzer()`.

rokopt commented on src/invariant/OrderBookIsNotCrossed.cpp:
```diff
@@ -0,0 +1,225 @@
+        return {};
+    }
+
+    if (lowestAsk < highestBid)
```
`lowestAsk == highestBid` and `lowestAsk < highestBid` are mutually exclusive, so how would we ever end up doing the (higher-complexity check) in the former path in a case that we could have avoided by doing the latter check first?

jonjove replied:
Yeah sorry. I was jumping around when I reviewed this and forgot that the previous check was gated on `lowestAsk == highestBid`.

rokopt commented on src/invariant/OrderBookIsNotCrossed.cpp:
```diff
@@ -0,0 +1,225 @@
+    {
+        auto existsEqualPriceNonPassive =
+            [](auto const targetPrice, Orders const& offers, auto& found) {
+                for (auto const& offer : offers)
```
The loop breaks as soon as it hits any offer whose price differs from that of the first offer in the list (`lowestAsk`), so it is taking advantage of the price-ordering, isn't it?

Within the same price, they're ordered by offer ID, i.e. the order in which they were created, but it's not apparent to me how to take advantage of that; if there were a bug that allowed some non-passive order to sneak into the midst of a long list of passive offers in the same `checkOnOperationApply()` call, all at the same price as the non-passive order, which should have crossed exactly equally with some pre-existing offer, then as far as I can tell we have to look at all the offers at that price to catch the bug.

jonjove replied:
My mistake, this does take advantage of the sort order.

rokopt commented on src/invariant/OrderBookIsNotCrossed.cpp:
```diff
@@ -0,0 +1,225 @@
+static std::string
+checkCrossed(Asset const& a, Asset const& b, OrderBook const& orderBook)
+{
+    // if either side of order book for asset pair empty or does not yet exist,
```
Will fix.

rokopt commented:
I've pushed changes to address all comments except on the floating-point reciprocal issue; these changes are small so I figure I may as well make them available for review while I think about the floating-point.

jonjove commented on src/invariant/OrderBookIsNotCrossed.h:
```diff
@@ -30,6 +30,14 @@ struct OfferEntryCmp
         {
             return true;
         }
+        else if (((a.flags & PASSIVE_FLAG) == 0) &&
```
This makes `OfferEntryCmp` violate the asymmetry requirement of https://en.cppreference.com/w/cpp/named_req/Compare.

Let `x < y` denote `OfferEntryCmp()(x, y)` for any `OfferEntry x, y`. If
```
L: Price = 1/1, Flags = PASSIVE_FLAG (offerID doesn't matter in this example)
R: Price = 2/1, Flags = 0 (offerID doesn't matter in this example)
```
Then

- `L < R == true` because `price(L) < price(R) == true`
- `R < L == true` because `price(R) < price(L) == false` and `(R.flags & PASSIVE_FLAG == 0) && (L.flags & PASSIVE_FLAG != 0) == true`

rokopt replied:
Thanks -- that test is supposed to be inside the prices-equal case, of course.  Sigh.  Fixed.

rokopt commented on src/invariant/OrderBookIsNotCrossed.cpp:
```diff
@@ -0,0 +1,225 @@
+
+    // check if crossed
+    auto lowestAsk = price(*asks.cbegin());
+    auto highestBid = 1 / price(*bids.cbegin());
```
One thing I am trying to confirm to myself -- or find a counterexample to -- as I consider this is whether we really can always resolve the question of whether the book is crossed by looking at the first offers in the per-asset-pair ask/bid lists.  I see this comment to `exchangeV10()`:

```
// exchangeV10 is a system for crossing offers that provides guarantees
// regarding the direction and magnitude of rounding errors:
// - When considering two crossing offers subject to a variety of limits,
//   exchangeV10 has a consistent approach to determining which offer is larger.
//   The smaller offer is always removed from the book.
// - When two offers cross, the rounding error will favor the offer that remains
//   in the book.
// - The rounding error will not favor either party by more than 1% (except in
//   the case of path payment, where the rounding error can favor the offer in
//   the book by an arbitrary amount). If the rounding error would exceed 1%,
//   no trade occurs and the smaller offer is removed.
```

In particular I think that the "smaller offer is removed" part means that rounding error should not allow the book to cross, but rather, if rounding error would prevent an exchange required to prevent the book from crossing, then one of the offers is simply removed altogether.  And I think that that means in turn that we indeed don't need to search beyond the first element:  I don't think there is a case where we are _expected_ to leave crossed offers in the book because of rounding errors preventing them from being taken.  But I have by no means absorbed all the rules yet.

rokopt replied:
> but suffice it to say that x = 1/49 is such an example

Bingo!  If I write this test:

```
    SECTION("Not crossed where highest bid = lowest ask but only because of "
            "passive offer(s), where double-precision reciprocal of price is "
            "not involutory")
    {
        auto const offer9 = createOffer(cur4, cur5, 1, Price{49, 1}, true);
        auto const offer10 = createOffer(cur5, cur4, 1, Price{1, 49}, true);
        applyCheck(*app, {offer9, offer10}, {}, true);
    }
```

Then I get:

```
2021-03-16T15:08:19.365 <test> [default FATAL] Order book is in a crossed state for { type = ASSET_TYPE_CREDIT_ALPHANUM4,
  alphaNum4 = {
    assetCode = 43555234,
    issuer = {
      type = PUBLIC_KEY_TYPE_ED25519,
      ed25519 = 0000000000000000000000000000000000000000000000000000000000000000
    }
  } }
 - { type = ASSET_TYPE_CREDIT_ALPHANUM4,
  alphaNum4 = {
    assetCode = 43555235,
    issuer = {
      type = PUBLIC_KEY_TYPE_ED25519,
      ed25519 = 0000000000000000000000000000000000000000000000000000000000000000
    }
  } }
 asset pair.
Top of the book is:
        Ask price: 49.0
        Bid price: 49.00000000000001

Where 49.00000000000001 >= 49.0!
```

rokopt replied:
As for fixing it...

`stellar-core` does use `double` when comparing prices, but as far as I can tell, it's in `isBetterOffer` where we're comparing prices with the same ask and bid assets, not with the reversed ones.  When we do a reciprocal, we do it in the rational form:

```
static Price
getInversePrice(Price const& price)
{
    return Price{price.d, price.n};
}
```

And when checking for crossing, we're using the `Price` comparison operators from `util/types.h`:

```
bool
ManageOfferOpFrameBase::doApply(AbstractLedgerTxn& ltxOuter)
{
[ ... ]
                if ((passive && (o.price >= maxWheatPrice)) ||
                    (o.price > maxWheatPrice))
                {
                    return OfferFilterResult::eStop;
                }
[ ... ]
}
```

Those operators use large rational arithmetic:

```
bool
operator>(Price const& a, Price const& b)
{
    uint128_t l(a.n);
    uint128_t r(a.d);
    l *= b.d;
    r *= b.n;
    return l > r;
}
```

So it's not obvious to me why we should be using `double` in the `OrderBookIsNotCrossed` invariant at all.  Indeed, if I switch to using the `Price` comparison operator, then the test above, and a similar one which should throw but doesn't when using `double`, both pass.  So I've committed those tests, and I've also committed changes to remove the use of `double` from the invariant module.

However, there are a couple things which I might want to change but that I'm not sure what the best way would be, so I'll ask here first:

- `types.h` only defines `>`, `>=`, and `==`; it doesn't define `<` or `<=`.  For now I just flipped the `<=` so that I wouldn't have to touch `types.h` in this PR, but I could add it if you prefer.
-  `types.h`'s version of `==` uses structural rather than rational equality:
```
bool
operator==(Price const& a, Price const& b)
{
    return (a.n == b.n) && (a.d == b.d);
}
```
So it's unsuitable for the uses in the invariant module.  In the committed code, I wrote something with the spelled-out name `pricesEqualAsRationals`, which works analogously to `>=` and `>`.  I'm guessing that some things are relying on the current semantics of `types.h`'s `==`, or else it wouldn't be that way, so I won't touch it unless someone can confirm otherwise.  I'm not sure whether I should do something more concise than `pricesEqualAsRationals` or whether the wordiness is actually good for clarity in this case.

jonjove commented on src/invariant/OrderBookIsNotCrossed.h:
```diff
@@ -30,16 +30,16 @@ struct OfferEntryCmp
-        }
         else if (priceA == priceB)
         {
+            if (((a.flags & PASSIVE_FLAG) == 0) &&
```
This still violates the asymmetry requirement.
```
L: Price = 1/1, Flags = 0, OfferID = 2
R: Price = 1/1, Flags = PASSIVE_FLAG, OfferID = 1
```

Then

- `L < R == true` because `price(L) == price(R)` and `(L.flags & PASSIVE_FLAG == 0) && (R.flags & PASSIVE_FLAG != 0) == true`
- `R < L == true` because `price(R) == price(L)` and `(R.flags & PASSIVE_FLAG == 0) && (L.flags & PASSIVE_FLAG != 0) == false` and `id(R) < id(L) == true`

rokopt replied:
Good catch -- I should be checking the reverse condition on flags as well.  I've pushed a commit to do that.

While staring at it for a while to see what _else_ I might have messed up, I also decided that I like it better without the else-after-returns.  So I removed them in a separate commit.  There appears to me to be some legitimate dispute over else-after-return, but `clang-tidy`, for example, is unambiguous (https://clang.llvm.org/extra/clang-tidy/checks/readability-else-after-return.html), as is Chromium (https://chromium.googlesource.com/chromium/src/+/refs/tags/73.0.3627.0/styleguide/c++/c++.md), and those two strike me as the most authoritative opinions that I've found so far.

rokopt replied:
I've also now added your two examples in a new `TEST_CASE` specifically for `OfferEntryCmp`.

jonjove commented:
@rokopt can you squash+rebase at this point? We've gone through a lot of iterations here, with considerable back-and-forth on changes, and I need to get a clean look at it.

rokopt commented:
> @rokopt can you squash+rebase at this point? We've gone through a lot of iterations here, with considerable back-and-forth on changes, and I need to get a clean look at it.

Sure, done.

jonjove commented on src/invariant/test/OrderBookIsNotCrossedTests.cpp:
```diff
@@ -0,0 +1,361 @@
+    }
+}
+
+TEST_CASE("OrderBookIsNotCrossed properly throws if order book is crossed",
```
One interesting group of cases that you aren't testing: changing the assets on an offer (this is permitted, you can even swap them). You might want to add some tests for this in both this case and the previous one.

rokopt replied:
That _is_ interesting.  I added several tests.  It also made me realize that you could cross the book by making a passive offer non-passive, so I added a test for that too.

rokopt commented on src/invariant/OrderBookIsNotCrossed.h:
```diff
@@ -0,0 +1,112 @@
+        }
+        if (priceA == priceB)
+        {
+            if (((a.flags & PASSIVE_FLAG) == 0) &&
```
Ah, that is nice.  Done.

rokopt commented on src/invariant/OrderBookIsNotCrossed.cpp:
```diff
@@ -0,0 +1,202 @@
+
+namespace stellar
+{
+double
```
I've removed uses of `static` in favor of either the anonymous namespace or of membership in `OrderBookNotCrossed`, depending on whether any members of the latter need the types/functions.

rokopt commented on src/invariant/OrderBookIsNotCrossed.cpp:
```diff
@@ -0,0 +1,202 @@
+
+namespace
+{
+static AssetPairSet
```
`static` removed.

rokopt commented on src/invariant/OrderBookIsNotCrossed.h:
```diff
@@ -0,0 +1,112 @@
+struct OfferEntryCmp
+{
+    bool
+    operator()(OfferEntry const& a, OfferEntry const& b) const
```
Agreed, I moved that into the cpp file.

rokopt commented on src/invariant/test/OrderBookIsNotCrossedTests.cpp:
```diff
@@ -0,0 +1,361 @@
+          "application behaviour",
+          "[invariant][OrderBookIsNotCrossed]")
+{
+    Asset cur1;
```
I factored that out to use `TestAccount::asset()`.

rokopt commented:
Note:  in addition to responding to the latest code review comments, I've also pushed a commit to remove the `mRollbackBeforeNextCheck` optimization, because I discovered that it causes occasional crashes during long fuzzing runs.  I don't understand why yet.  We can either delay this PR until I've found a fix, or merge it and file an issue to perform that optimization (correctly) afterwards, as you prefer.

jonjove commented on src/invariant/test/OrderBookIsNotCrossedTests.cpp:
```diff
@@ -1,3 +1,4 @@
+#include "test/TestAccount.h"
```
Nitpick: this looks like it's in the wrong spot?

rokopt replied:
Indeed.  And odd.  Fixed.

rokopt commented:
> Note: in addition to responding to the latest code review comments, I've also pushed a commit to remove the `mRollbackBeforeNextCheck` optimization, because I discovered that it causes occasional crashes during long fuzzing runs. I don't understand why yet. We can either delay this PR until I've found a fix, or merge it and file an issue to perform that optimization (correctly) afterwards, as you prefer.

This optimization is restored now, in a corrected form.

MonsieurNicolas commented:
Looks like it's basically ready to go? Needs to be rebased and we can merge. @jonjove what do you think?

As a sanity check I didn't see an update on the fuzzer stability front (only one I see here is the one from very early on). So maybe @rokopt can post this in parallel?

jonjove commented:
@MonsieurNicolas I just went through and resolved all open comments from me. Should be ready for rebase/squash/merge. Agreed that I'd like to see the fuzzer statistics though.

rokopt commented:
I've squashed and rebased.  Stability didn't change noticeably in this PR; it's still at the ~92% where it's been since the ledger-key mapping work:

![afl-stats-with-squashed-order-book-not-crossed-pr](https://user-images.githubusercontent.com/17652016/112560100-a996bc80-8d8f-11eb-95be-2ad4a75c55c8.png)

MonsieurNicolas commented:
mmm, I do have a question though. Is the "exec speed" really that slow compared to both master and the previous iteration of that PR?

I would not expect such a difference as the vast majority of fuzzing should not involve this invariant (it is, after all, unlikely that the fuzzer explores branches with offers).

MonsieurNicolas commented on src/invariant/OrderBookIsNotCrossed.cpp:
```diff
@@ -0,0 +1,230 @@
+{
+    updateOrderBook(ltxDelta);
+    auto assetPairs = extractAssetPairs(ltxDelta);
+    return assetPairs.size() > 0 ? check(assetPairs) : std::string{};
```
I *think* that the reason things are super slow is that `assetPairs.size() > 0` is always true even when there are no changes to the order book from this operation. In fact we should not perform any work in `extractAssetPairs` if the operation didn't touch the order book.

rokopt replied:
Hmm, why must `assetPairs.size()` be `> 0`?  It looks to me as though if there are no `OFFER`s in `LedgerTxnDelta`, then it will be `0`, and I put in some logs and they say that it is often `0`.

One thing I mentioned to Jon about performance which I should bring up is that exec speed of fuzzing seems highly variable on my laptop, even from run to run without code changes (unlike stability, for example, which is generally within a fraction of a percent across many runs).  Fuzzing does make my laptop fan roar, and I have a suspicion (though no confirmation) that there might be highly variable amounts of CPU-throttling going on depending on I-don't-know-what thermal factors.

MonsieurNicolas replied:
meh and I thought my head was straight in the morning :)

yeah this is extracting only from the current txn object so not it

MonsieurNicolas commented:
I think I found the bug, let me know if this makes sense @rokopt

MonsieurNicolas commented:
I ran the version with the invariant enabled and without based on the same initial corpus (so no variation because of that) and they performed the same way, so it was probably a fluke with your computer

MonsieurNicolas commented:
we'll merge as soon as master reopens
